File: internal/database/labels.go
package database

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// CreateLabel creates a new label for a user
func (db *DB) CreateLabel(ctx context.Context, label *Label) error {
	// Convert reasons slice to JSON
	reasonsJSON, err := json.Marshal(label.Reasons)
	if err != nil {
		return fmt.Errorf("failed to marshal reasons: %w", err)
	}

	query := `
		INSERT INTO labels (user_id, name, reasons, description, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id
	`

	err = db.conn.QueryRowContext(
		ctx,
		query,
		label.UserID,
		label.Name,
		reasonsJSON,
		label.Description,
		time.Now(),
		time.Now(),
	).Scan(&label.ID)

	if err != nil {
		return fmt.Errorf("failed to create label: %w", err)
	}

	return nil
}

// GetAllLabels retrieves all labels for a user
func (db *DB) GetAllLabels(ctx context.Context, userID int64) ([]*Label, error) {
	query := `
		SELECT id, user_id, name, reasons, description, created_at, updated_at
		FROM labels
		WHERE user_id = $1
		ORDER BY name
	`

	rows, err := db.conn.QueryContext(ctx, query, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to query labels: %w", err)
	}
	defer rows.Close()

	var labels []*Label
	for rows.Next() {
		var label Label
		var reasonsJSON []byte

		err := rows.Scan(
			&label.ID,
			&label.UserID,
			&label.Name,
			&reasonsJSON,
			&label.Description,
			&label.CreatedAt,
			&label.UpdatedAt,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to scan label: %w", err)
		}

		// Unmarshal reasons JSON
		if err := json.Unmarshal(reasonsJSON, &label.Reasons); err != nil {
			return nil, fmt.Errorf("failed to unmarshal reasons: %w", err)
		}

		labels = append(labels, &label)
	}

	if err = rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating labels: %w", err)
	}

	return labels, nil
}

// UpdateLabel updates a label's name, description, and reasons
func (db *DB) UpdateLabel(ctx context.Context, label *Label) error {
	reasonsJSON, err := json.Marshal(label.Reasons)
	if err != nil {
		return fmt.Errorf("failed to marshal reasons: %w", err)
	}

	query := `
		UPDATE labels SET name = $1, description = $2, reasons = $3, updated_at = $4
		WHERE id = $5 AND user_id = $6
	`

	result, err := db.conn.ExecContext(ctx, query, label.Name, label.Description, reasonsJSON, time.Now(), label.ID, label.UserID)
	if err != nil {
		return fmt.Errorf("failed to update label: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return fmt.Errorf("label not found or not owned by user")
	}

	return nil
}

// DeleteLabel deletes a label for a user
func (db *DB) DeleteLabel(ctx context.Context, userID int64, labelID string) error {
	query := `DELETE FROM labels WHERE id = $1 AND user_id = $2`

	result, err := db.conn.ExecContext(ctx, query, labelID, userID)
	if err != nil {
		return fmt.Errorf("failed to delete label: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return fmt.Errorf("label not found or not owned by user")
	}

	return nil
}

File: cloudflare/src/db/labels.ts
import type { Label, LabelRow } from '../types/models';

function mapLabel(row: LabelRow): Label {
  let reasons: string[] = [];
  try {
    reasons = JSON.parse(row.reasons) as string[];
  } catch {
    reasons = [];
  }
  return {
    id: row.id,
    userId: row.user_id,
    name: row.name,
    reasons,
    description: row.description ?? '',
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export async function createLabel(db: D1Database, label: Omit<Label, 'id' | 'createdAt' | 'updatedAt'>): Promise<number> {
  const now = new Date().toISOString();
  const row = await db
    .prepare(
      `INSERT INTO labels (user_id, name, reasons, description, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?)
       RETURNING id`,
    )
    .bind(label.userId, label.name, JSON.stringify(label.reasons), label.description, now, now)
    .first<{ id: number }>();

  if (!row) throw new Error('Failed to create label');
  return row.id;
}

export async function getAllLabels(db: D1Database, userId: number): Promise<Label[]> {
  const { results } = await db
    .prepare(
      `SELECT id, user_id, name, reasons, description, created_at, updated_at
       FROM labels
       WHERE user_id = ?
       ORDER BY name`,
    )
    .bind(userId)
    .all<LabelRow>();
  return results.map(mapLabel);
}

export async function getUserLabelsWithDetails(db: D1Database, userId: number): Promise<Label[]> {
  const { results } = await db
    .prepare(
      `SELECT id, user_id, name, COALESCE(description, '') as description, reasons, created_at, updated_at
       FROM labels
       WHERE user_id = ?
       ORDER BY name`,
    )
    .bind(userId)
    .all<LabelRow>();
  return results.map(mapLabel);
}

export async function updateLabel(db: D1Database, label: Label): Promise<void> {
  const now = new Date().toISOString();
  const result = await db
    .prepare(
      `UPDATE labels SET name = ?, description = ?, reasons = ?, updated_at = ?
       WHERE id = ? AND user_id = ?`,
    )
    .bind(label.name, label.description, JSON.stringify(label.reasons), now, label.id, label.userId)
    .run();

  if (result.meta.changes === 0) {
    throw new Error('label not found or not owned by user');
  }
}

export async function deleteLabel(db: D1Database, userId: number, labelId: number): Promise<void> {
  const result = await db
    .prepare('DELETE FROM labels WHERE id = ? AND user_id = ?')
    .bind(labelId, userId)
    .run();

  if (result.meta.changes === 0) {
    throw new Error('label not found or not owned by user');
  }
}
